module_utils.(eos, nxos) - Support use_proxy argument

DamianZaremba opened the pull request:
##### SUMMARY
Running ansible with a proxy set in the environment causes the eos module to
attempt to connect to devices via the proxy.

To prevent this behaviour the only way is to unset the proxy out of the
environment, either by wrapping the ansible calls or doing it in a piece of code
executed before connect, such as a vars_module (though this is very hacky).

This change allows you to set `use_proxy: no` under the provider config.

The default value is set to True, which mirrors the behaviour seen today.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
module_utils/eos.py
modules/network/eos/*.py

##### ANSIBLE VERSION
```
ansible 2.4.0.0
  config file = None
  configured module search path = ['/home/dzaremba/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /opt/blue-python/3.6/lib/python3.6/site-packages/ansible
  executable location = /bin/ansible
  python version = 3.6.2 (default, Jul 18 2017, 11:34:24) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
```
(This is a straight package of the upstream release)

##### ADDITIONAL INFORMATION
Executing the below eos_config task without this change and https_proxy set in your environment (to a proxy that can't connect to the target) the execution fails with `Request failed: <urlopen error Tunnel connection failed: 503 Service Unavailable>`

```
    - name: Test
      eos_config:
        src: "{{ device_conf }}"
        replace: config
        provider:
          transport: eapi
      check_mode: yes
```

With this change and the below eos_config task, the task is executed as expected, directly connecting to the device.

```
    - name: Test
      eos_config:
        src: "{{ device_conf }}"
        replace: config
        provider:
          transport: eapi
          use_proxy: no
      check_mode: yes
```

DamianZaremba committed:
eos - Support use_proxy argument

Running ansible with a proxy set in the environment causes the eos module to
attempt to connect to devices via the proxy.

To prevent this behaviour the only way is to unset the proxy out of the
environment, either by wrapping the ansible calls or doing it in a piece of code
executed before connect, such as a vars_module (though this is very hacky).

This change allows you to set `use_proxy: no` under the provider config.

The default value is set to True, which mirrors the behaviour seen today.

ganeshrn commented on lib/ansible/module_utils/eos.py:
```diff
@@ -50,6 +50,7 @@
     'auth_pass': dict(no_log=True, fallback=(env_fallback, ['ANSIBLE_NET_AUTH_PASS'])),
 
     'use_ssl': dict(default=True, type='bool'),
+    'use_proxy': dict(default=True, type='bool'),
```
Can you please add documentation for `use_proxy` [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/utils/module_docs_fragments/eos.py)

ganeshrn reviewed (changes requested):
Please add doc for `use_proxy`

DamianZaremba commented:
Docs added

gundalow commented on lib/ansible/utils/module_docs_fragments/eos.py:
```diff
@@ -120,6 +120,11 @@ class ModuleDocFragment(object):
+        description:
+          - If C(no), proxy environment variables (http_proxy/https_proxy) will be ignored.
+        default: 'yes'
+        choices: ['yes', 'no']
```
Needs a `version_added: "2.5"`

gundalow commented:
I wonder if we should do the same for NXOS's nxapi as well?

abadger commented:
Is this change to the module better or worse than changing the http_proxy type environment variables via ansible's environment: keyword?

http://docs.ansible.com/ansible/latest/playbooks_environment.html

gundalow commented on lib/ansible/utils/module_docs_fragments/eos.py:
```diff
@@ -120,6 +120,11 @@ class ModuleDocFragment(object):
         choices: ['yes', 'no']
+      use_proxy:
+        description:
+          - If C(no), proxy environment variables (http_proxy/https_proxy) will be ignored.
```
`If C(no), the environment variables C(http_proxy) and C(https_proxy) will be ignored.

gundalow commented:
@Qalthos Does this change make sense for `2.5.0` given the connection work you will be doing?

Also as @abadger points out, is this "module specific", should we be supporting the `environment:` keyword to be in line with how Ansible works configuring Linux systems?

Qalthos commented:
At the moment, the only connection work that is definitely getting in 2.5 is ssh-based... HTTP(S) apis might make it, but I see no reason this would impact that work.

gundalow commented:
@DamianZaremba Can you please rebase and add the same for nxos then we can include this in Ansible 2.5.0. Thanks in advance

DamianZaremba committed:
nexos - Support use_proxy argument

Running ansible with a proxy set in the environment causes the nexos module to
attempt to connect to devices via the proxy.

To prevent this behaviour the only way is to unset the proxy out of the
environment, either by wrapping the ansible calls or doing it in a piece of code
executed before connect, such as a vars_module (though this is very hacky).

This change allows you to set `use_proxy: no` under the provider config.

The default value is set to True, which mirrors the behaviour seen today.

DamianZaremba commented:
I've squashed the eos changes into 1 commit and added the code for nxos in a second (I don't have a NX-OS box to hand to test it... but it's pretty simple). Let me know if you'd like any more changes.

gundalow commented:
Changes look good, though I have one implementation question.
Should the default be enabled or disabled?

Having this disabled by default would match current (Ansible 2.4) behavior.

DamianZaremba commented:
Proxy usage is currently enabled by default on fetch_url (https://github.com/ansible/ansible/blob/stable-2.4/lib/ansible/module_utils/urls.py#L992), so the default option for `use_proxy` being True here should carry through the default as it stands now (unless you specify a value).

Flipping `use_proxy` to False by default, would reverse the logic as it is now... it's a little confusing since we're exposing a default, rather than adding an option really (if Ansible should use system proxies by default or not is a wider question, that isn't specific to these usages I don't think).

I might have totally misunderstood your point though.

gundalow commented:
@DamianZaremba Your points are correct. I hadn't thought about fetch_url.

I was wondering if we were changing behavior for eos users. Though thinking about this a bit more I guess they wouldn't have the `environment` keyword set.

So I'm +1 on this

gundalow commented:
@mikewiebe @trishnaguha @Qalthos  NXOS change

gundalow commented:
oh, we should document this on http://docs.ansible.com/ansible/latest/network_debug_troubleshooting.html#proxy-issues

gundalow commented:
Will merge this and do the documentation later (tracked in https://github.com/ansible/ansible/issues/34697)

Logic is this change needs to be in before the 2.5.0 freeze, though docs can come a bit later.

@DamianZaremba Thanks again for doing this

gundalow commented:
Merged into devel, will be released in `2.5.0`
